cpp/src/cider/util/CiderParquetReader.cpp

```diff
@@ -0,0 +1,437 @@
+  for (int i = 0; i < totalRowGroups_; i++) {
+    rowGroupReaders_[i] = parquetReader_->RowGroup(firstRowGroupIndex_ + i);
+    totalRows_ += rowGroupReaders_[i]->metadata()->num_rows();
+    LOG(INFO) << "this rg have rows: " << rowGroupReaders_[i]->metadata()->num_rows();
+  }
+}
+
+bool Reader::checkEndOfRowGroup() {
+  if (totalRowsRead_ != totalRowsLoadedSoFar_) {
+    return false;
+  }
+
+  // rowGroupReaders index starts from 0
+  rowGroupReader_ = rowGroupReaders_[currentRowGroup_ - firstRowGroupIndex_];
+  currentRowGroup_++;
+  totalRowGroupsRead_++;
+
+  for (int i = 0; i < requiredColumnIndex_.size(); i++) {
+    columnReaders_[i] = rowGroupReader_->Column(requiredColumnIndex_[i]);
+  }
+
+  totalRowsLoadedSoFar_ += rowGroupReader_->metadata()->num_rows();
+  LOG(INFO) << "totalRowsLoadedSoFar: " << totalRowsLoadedSoFar_;
+  return true;
+}
+
+void Reader::allocateBuffers(int rowsToRead,
+                             std::vector<int64_t*>& buffersPtr,
+                             std::vector<uint8_t*>& nullsPtr) {
+  for (int i = 0; i < buffersPtr.size(); i++) {
+    switch (parquetTypeVector_[i]) {
+      case parquet::Type::BOOLEAN:
+        buffersPtr[i] = (int64_t*)allocator_->allocate(rowsToRead);
+      case parquet::Type::INT32:
+      case parquet::Type::FLOAT:
+        buffersPtr[i] = (int64_t*)allocator_->allocate(rowsToRead * 4);
+        break;
+      case parquet::Type::INT64:
+      case parquet::Type::DOUBLE:
+        buffersPtr[i] = (int64_t*)allocator_->allocate(rowsToRead * 8);
+        break;
+      case parquet::Type::INT96:
+        buffersPtr[i] =
+            (int64_t*)allocator_->allocate(rowsToRead * sizeof(parquet::Int96));
+        break;
+      case parquet::Type::BYTE_ARRAY:
+        buffersPtr[i] =
+            (int64_t*)allocator_->allocate(rowsToRead * sizeof(parquet::ByteArray));
+        break;
+      default:
+        CIDER_THROW(CiderCompileException, "unsupport type");
+    }
+    nullsPtr[i] = (uint8_t*)allocator_->allocate((rowsToRead + 7) >> 3);
+  }
+}
+
+// return rows actually read
+int Reader::readBatch(int32_t batchSize,
+                      ArrowSchema*& outputSchema,
+                      ArrowArray*& outputArray) {
+  // this reader have read all rows
+  if (totalRowsRead_ >= totalRows_) {
+    return -1;
+  }
+
+  // at most read to the end of current row group
+  checkEndOfRowGroup();
+
+  int rowsToRead = std::min((int64_t)batchSize, totalRowsLoadedSoFar_ - totalRowsRead_);
+
+  std::vector<int64_t*> buffersPtr(requiredColumnNum_);
+  std::vector<uint8_t*> nullsPtr(requiredColumnNum_);
+
+  allocateBuffers(rowsToRead, buffersPtr, nullsPtr);
+
+  currentBatchSize_ = batchSize;
+
+  int rowsActualRead = doReadBatch(rowsToRead, buffersPtr, nullsPtr);
+
+  totalRowsRead_ += rowsActualRead;
+  LOG(INFO) << "total rows read yet: " << totalRowsRead_;
+  LOG(INFO) << "ret rows " << rowsActualRead;
+
+  auto schema_and_array = convert2Arrow(rowsToRead, buffersPtr, nullsPtr);
```

Comment:
if read to arrow why not use parquet arrow api to read the column to chunkarry then convert to arrowarray by arrow bridge

Reply:
we need to call `arrow::Table::CombineChunksToBatch()` first to convert to `RecordBatch`, and then call `ExportRecordBatch()`. the first step may have extra memory copy.